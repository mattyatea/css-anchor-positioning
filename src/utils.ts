import {
  type Atrule,
  type CssNode,
  type Declaration,
  type FunctionNode,
  type Identifier,
  List,
  type Selector as CssTreeSelector,
  type SelectorList,
  type StyleSheet,
  type Value,
} from 'css-tree';
import generate from 'css-tree/generator';
import parse from 'css-tree/parser';
import { clone } from 'css-tree/utils';
import { nanoid } from 'nanoid/non-secure';

import type { Selector } from './dom.js';

export const INSTANCE_UUID = nanoid();

// https://github.com/import-js/eslint-plugin-import/issues/3019

export interface DeclarationWithValue extends Declaration {
  value: Value;
}

export function isAnchorFunction(node: CssNode | null): node is FunctionNode {
  return Boolean(node && node.type === 'Function' && node.name === 'anchor');
}

export function getAST(cssText: string, parseAtrulePrelude = false) {
  return parse(cssText, {
    parseAtrulePrelude: parseAtrulePrelude,
    parseCustomProperty: true,
  });
}

export function generateCSS(ast: CssNode) {
  return generate(ast, {
    // Default `safe` adds extra (potentially breaking) spaces for compatibility
    // with old browsers.
    mode: 'spec',
  });
}

export function isDeclaration(node: CssNode): node is DeclarationWithValue {
  return node.type === 'Declaration';
}

export function getDeclarationValue(node: DeclarationWithValue) {
  return (node.value.children.first as Identifier).name;
}

function isImportRule(node: CssNode): node is Atrule {
  return node.type === 'Atrule' && node.name === 'import';
}

export function makeImportUrlAbsolute(node: CssNode): boolean {
  if (!isImportRule(node)) return false;
  // AtRulePreludes are not parsed by default, so we need to reparse the node
  // to get the URL.
  const reparsedNode = (getAST(generateCSS(node), true) as StyleSheet).children
    .first;
  if (!reparsedNode || !isImportRule(reparsedNode)) return false;
  if (reparsedNode.prelude?.type !== 'AtrulePrelude') return false;

  // URL is always the first child of the prelude
  const url = reparsedNode.prelude.children.first;
  if (!url) return false;
  if (url.type !== 'Url' && url.type !== 'String') return false;
  url.value = new URL(url.value, document.baseURI).href;
  node.prelude = reparsedNode.prelude;
  return true;
}

export function makeDeclarationValueUrlAbsolute(node: CssNode): boolean {
  if (!isDeclaration(node)) return false;

  if (node.value.type !== 'Value') return false;
  let changed = false;

  const mapped = node.value.children.toArray().map((child) => {
    if (!child) return child;
    if (child.type !== 'Url') return child;
    child.value = new URL(child.value, document.baseURI).href;
    changed = true;
    return child;
  });
  node.value.children = new List<CssNode>().fromArray(mapped);
  return changed;
}

export interface StyleData {
  el: HTMLElement;
  css: string;
  url?: URL;
  changed?: boolean;
  created?: boolean; // Whether the element is created by the polyfill
}

export const POSITION_ANCHOR_PROPERTY = `--position-anchor-${INSTANCE_UUID}`;

export function splitCommaList(list: List<CssNode>) {
  return list.toArray().reduce(
    (acc: Identifier[][], child) => {
      if (child.type === 'Operator' && child.value === ',') {
        acc.push([]);
        return acc;
      }
      if (child.type === 'Identifier') {
        acc[acc.length - 1].push(child);
      }

      return acc;
    },
    [[]],
  );
}

export function getSelectors(rule: SelectorList | undefined) {
  if (!rule) return [];

  return (rule.children as List<CssTreeSelector>)
    .map((selector) => {
      let pseudoElementPart: string | undefined;

      if (selector.children.last?.type === 'PseudoElementSelector') {
        selector = clone(selector) as CssTreeSelector;
        pseudoElementPart = generateCSS(selector.children.last!);
        selector.children.pop();
      }

      const elementPart = generateCSS(selector);

      return {
        selector: elementPart + (pseudoElementPart ?? ''),
        elementPart,
        pseudoElementPart,
      } satisfies Selector;
    })
    .toArray();
}
