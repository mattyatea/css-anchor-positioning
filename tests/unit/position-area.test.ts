import type { Rule, StyleSheet } from 'css-tree';

import {
  axisForPositionAreaValue,
  parsePositionAreaValue,
} from '../../src/position-area.js';
import { getAST } from '../../src/utils.js';

const createBlock = (input = 'a{b:c}') => {
  const css = getAST(input) as StyleSheet;
  return (css.children.first! as Rule).block;
};

describe('position-area', () => {
  afterAll(() => {
    document.head.innerHTML = '';
    document.body.innerHTML = '';
  });

  describe('axisForPositionAreaValue', () => {
    it.each([
      ['block-start', 'block'],
      ['block-end', 'block'],
      ['top', 'block'],
      ['span-top', 'block'],
      ['span-self-block-end', 'block'],
      ['left', 'inline'],
      ['span-right', 'inline'],
      ['span-x-self-start', 'inline'],
      ['center', 'ambiguous'],
      ['span-all', 'ambiguous'],
      ['start', 'ambiguous'],
      ['end', 'ambiguous'],
      ['span-end', 'ambiguous'],
      ['self-span-start', 'ambiguous'],
    ])('%s as %s', (input, expected) => {
      expect(axisForPositionAreaValue(input)).toBe(expected);
    });
  });
  describe('parsePositionAreaValue', () => {
    // Valid cases
    it.each([
      [['left', 'bottom'], { block: 'bottom', inline: 'left' }],
      [['bottom', 'left'], { block: 'bottom', inline: 'left' }],
      [['x-start', 'y-end'], { block: 'y-end', inline: 'x-start' }],
    ])('%s parses', (input, expected) => {
      expect(
        parsePositionAreaValue(input, createBlock())?.values,
      ).toMatchObject(expected);
    });

    // With ambiguous values
    it.each([
      [['left', 'center'], { block: 'center', inline: 'left' }],
      [['center', 'left'], { block: 'center', inline: 'left' }],
      [['span-all', 'y-end'], { block: 'y-end', inline: 'span-all' }],
    ])('%s parses values', (input, expected) => {
      expect(
        parsePositionAreaValue(input, createBlock())?.values,
      ).toMatchObject(expected);
    });
    it.each([
      [['left', 'center'], { block: [1, 2], inline: [0, 1] }],
      [['center', 'left'], { block: [1, 2], inline: [0, 1] }],
      [['span-all', 'y-end'], { block: [2, 3], inline: [0, 3] }],
    ])('%s parses grid', (input, expected) => {
      expect(parsePositionAreaValue(input, createBlock())?.grid).toMatchObject(
        expected,
      );
    });

    // With single values
    it.each([
      [['top'], { block: 'top', inline: 'span-all' }],
      [['center'], { block: 'center', inline: 'center' }],
      [['start'], { block: 'start', inline: 'start' }],
      [['self-start'], { block: 'self-start', inline: 'self-start' }],
      [['span-end'], { block: 'span-end', inline: 'span-end' }],
    ])('%s parses', (input, expected) => {
      expect(
        parsePositionAreaValue(input, createBlock())?.values,
      ).toMatchObject(expected);
    });

    // Invalid, can't parse
    it.each([[['left', 'left']], [['left', 'block-end']]])(
      '%s is undefined',
      (input) => {
        expect(parsePositionAreaValue(input, createBlock())).toEqual(undefined);
      },
    );
  });

  describe('insets', () => {
    it.each([
      [
        ['top', 'right'],
        [0, 'ANCHOR_TOP'],
        ['ANCHOR_RIGHT', 0],
      ],
      [
        ['bottom', 'left'],
        ['ANCHOR_BOTTOM', 0],
        [0, 'ANCHOR_LEFT'],
      ],
      [
        ['center', 'center'],
        ['ANCHOR_TOP', 'ANCHOR_BOTTOM'],
        ['ANCHOR_LEFT', 'ANCHOR_RIGHT'],
      ],
    ])('%s', (input, block, inline) => {
      const res = parsePositionAreaValue(input, createBlock())!.insets;
      expect(res.block).toEqual(block);
      expect(res.inline).toEqual(inline);
    });
  });

  describe('axisAlignment', () => {
    it.each([
      [['top', 'right'], 'end', 'start'],
      [['bottom', 'left'], 'start', 'end'],
      [['center', 'center'], 'center', 'center'],
    ])('%s', (input, block, inline) => {
      const res = parsePositionAreaValue(input, createBlock())!.alignments;
      expect(res.block).toEqual(block);
      expect(res.inline).toEqual(inline);
    });
  });
});
