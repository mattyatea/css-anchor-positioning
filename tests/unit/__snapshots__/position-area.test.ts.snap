// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`position-area > activeWrapperStyles > returns the active styles 1`] = `"    [data-anchor-position-wrapper="selectorUUID"][data-pa-wrapper-for-targetUUID] {      --pa-value-top: var(targetUUID-top);      --pa-value-left: var(targetUUID-left);      --pa-value-right: var(targetUUID-right);      --pa-value-bottom: var(targetUUID-bottom);      --pa-value-justify-self: var(targetUUID-justify-self);      --pa-value-align-self: var(targetUUID-align-self);    }  "`;
